### ISSUE TITLE
tf.truncated_normal_initializer missing

### ISSUE THREAD

**Issue by Hondo2019** [losttech/Gradient#8]
Hi there! Sometimes i migrate code from a to b and today i tried to convert this: https://github.com/dongjun-Lee/text-classification-models-tf/blob/master/cnn_models/vd_cnn.py to c# using your gradient. 

Not sure if this will ever work, acually not even the original py seems to work on windows lol (training just stops after some time). 

With gradient there are some compile problems:
- tf.truncated_normal_initializer seems to be missing
- tf.layers.conv2d requires more parameters than the original py, not sure what to pass to these additional parameters like strides, padding, data_format, dilation_rate
- tf.layers.max_pooling2d.strides requires a tuple with 3 ints, py needs just 2 ints. 

Any ideas?

**Comment by lostmsu** [comment]
> `tf.truncated_normal_initializer`

It is a class in `tensorflow` namespace. E.g. just do `new truncated_normal_initializer(...)` or `truncated_normal_initializer.NewDyn(...)`.

> tf.layers.conv2d requires more parameters than the original py
> tf.layers.max_pooling2d.strides requires a tuple with 3 ints, py needs just 2 ints.

In cases like this just use `tf.layers.conv2d_dyn`. See [Dynamic Overloads](https://github.com/losttech/Gradient#dynamic-overloads).

**Comment by lostmsu** [comment]
I'd also encourage you to read the relevant section of Gradient documentation: [Porting Python code to Gradient](https://github.com/losttech/Gradient#porting-python-code-to-gradient-c)

**Comment by Hondo2019** [comment]
Thanks to your help I got it building now. Now facing some runtime errors. 

py looks like this: 
`
         with tf.name_scope("loss"):
            y_one_hot = tf.one_hot(self.y, num_class)
            self.loss = tf.reduce_mean(
                tf.nn.softmax_cross_entropy_with_logits_v2(logits=self.logits, labels=y_one_hot))
`

with gradient I did: 
`
            var y_one_hot = tf.one_hot(y, num_class);
            var loss = tf.reduce_mean(
                tf.nn.softmax_cross_entropy_with_logits_v2(logits: logits, labels: y_one_hot));
`
that throws 'softmax_cross_entropy_with_logits_v2()  .. . has some invalid arguments'

I than tried:
`
            var y_one_hot = tf.one_hot(y, num_class);
            var loss = tf.reduce_mean(
                tf.nn.softmax_cross_entropy_with_logits_v2_dyn(logits: logits, labels: y_one_hot));
`
which throws 'softmax_cross_entropy_with_logits_v2() takes 2 to 4 positional arguments but 5 were given'

logits and y_one_hot are both of type Tensor.

**Comment by lostmsu** [comment]
@Hondo2019 are you using TensorFlow 2.0? In the docs for https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits_v2 I see 1.14 still has 5 arguments.

I'll try to see what you can do to work around that though if you want to continue using the version you are using now.

**Comment by Hondo2019** [comment]
I am fine with any version. I guess I am using 1.14 now. The stuff I try to migrate (https://github.com/dongjun-Lee/text-classification-models-tf/blob/master/cnn_models/vd_cnn.py) passes 2 parameters to softmax_cross_entropy_with_logits_v2 (labels & logits) and the other three should be optional, right?

**Comment by Hondo2019** [comment]
The remaining three parameters are just axis, name and dim. I will check back and try with axis: -1 as soon as I have the chance. Maybe that helps.

**Comment by lostmsu** [comment]
@Hondo2019 well, that's a technical detail how Gradient works. In .NET the parameters are always passed. If you don't do it, compiler just substitutes the default value.
E.g. `int X(int v = 0) { ... }` followed by `X()` is compiled as `X(0)`.

**Comment by Hondo2019** [comment]
Thats true, this wont help than. 

If its says it just wants 2 to 4 parameters, seems like I use a version of tensorflow where the depricated dim is gone already. Which should not be the case in 1.14 as you mentioned. Ok I will check again, which version of TF I am using.

**Comment by lostmsu** [comment]
OK, verified, that `((dynamic)tf.nn.softmax_cross_entropy_with_logits_v2_fn)(logits: logits, labels: y_one_hot)` works with 2.0-rc0.

But at least for now while Gradient is in preview, I recommend using 1.10 to avoid the need for this, unless you require specific features from the later versions.

I actually used 1.13 quite a lot, so I suspect 1.1x would be mostly OK.

**Comment by Hondo2019** [comment]
I can confirm now that my first tests were with tensorflow 1.14.0. I than tried to use 1.13.1 and 1.13.2 but the softmax error message did not change. Older versions do not seem to be available any more (I am using python 3.7.3). v2.0.0rc0 does not seem to work at all with gradient right now, it crashes at the very first codeline where its trying to get the cnn initializer instance. 

Do you think its possible to make the softmax_cross_entropy_with_logits_v2_fn work with one of the currently available tensorflow versions?

**Comment by lostmsu** [comment]
@Hondo2019 have you tried `((dynamic)tf.nn.softmax_cross_entropy_with_logits_v2_fn)(logits: logits, labels: y_one_hot)`?

**Comment by Hondo2019** [comment]
ups, did not notice the _fn, sorry. 

Yes, it really works with tf.nn.softmax_cross_entropy_with_logits_v2_fn, but than we have the same thing with tf.reduce_mean. 

The code is:
```
            var loss = tf.reduce_mean(input_tensor: 
                ((dynamic)tf.nn.softmax_cross_entropy_with_logits_v2_fn)(logits: logits, labels: y_one_hot));
```

it says: 'takes 1 to 4 positional arguments but 6 were given'. Same with _dyn, did not get it building with _fn.

**Comment by lostmsu** [comment]
@Hondo2019 looks like the 1.14 approach is way too problematic.
Can you use Python 3.6? I believe you can't install 1.10.x because it is not available for 3.7. See https://pypi.org/project/tensorflow/1.10.1/

**Comment by lostmsu** [comment]
The upcoming preview will support v1.15

**Comment by lostmsu** [comment]
Preview 7 is released with TensorFlow 1.15 support
